exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,490 @@
+    description: Option<String>,
+}
+
+impl TransactionError {
+    /// Creates a new `TransactionError` instance with the specified error type and description.
+    fn new(error_type: TransactionErrorType, description: Option<String>) -> Self {
+        Self {
+            error_type,
+            description,
+        }
+    }
+
+    /// Creates a new `TransactionError` instance with the specified error code and description.
+    fn code(code: u8, description: Option<String>) -> Self {
+        Self::new(TransactionErrorType::Code(code), description)
+    }
+
+    /// Creates a new `TransactionError` representing panic with the given description.
+    fn panic(description: Option<String>) -> Self {
+        Self::new(TransactionErrorType::Panic, description)
+    }
+
+    /// Creates a new `TransactionError` instance from `std::thread::Result`'s `Err`.
+    pub(crate) fn from_panic(_panic: &Box<Any + Send>) -> Self {
+        // TODO: Try to get description from panic.
+        Self::panic(None)
+    }
+
+    /// Returns error type of this `TransactionError` instance.
+    pub fn error_type(&self) -> TransactionErrorType {
+        self.error_type
+    }
+
+    /// Returns an optional error description.
+    pub fn description(&self) -> &Option<String> {
+        &self.description
+    }
+}
+
+impl<'a, T: Transaction> From<T> for Box<Transaction + 'a> {
+    fn from(tx: T) -> Self {
+        Box::new(tx) as Box<Transaction>
+    }
+}
+
+impl fmt::Display for TransactionError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.error_type {
+            TransactionErrorType::Panic => write!(f, "Panic during execution")?,
+            TransactionErrorType::Code(c) => write!(f, "Error code: {}", c)?,
+        }
+
+        if let Some(ref description) = self.description {
+            write!(f, " description: {}", description)?;
+        }
+
+        Ok(())
+    }
+}
+
+// String content (`TransactionError::Description`) is intentionally excluded from the hash
+// calculation because user can be tempted to use error description from a third-party libraries
+// which aren't stable across the versions.
+impl CryptoHash for TransactionResult {
+    fn hash(&self) -> Hash {
+        u16::hash(&status_as_u16(self))
+    }
+}
+
+impl From<ExecutionError> for TransactionError {
+    fn from(error: ExecutionError) -> Self {
+        Self {
+            error_type: TransactionErrorType::Code(error.code),
+            description: error.description,
+        }
+    }
+}
+
+// `TransactionResult` is stored as `u16` plus `bool` (`true` means that optional part is present)
+// with optional string part needed only for string error description.
+impl StorageValue for TransactionResult {
+    fn into_bytes(self) -> Vec<u8> {
+        let mut res = u16::into_bytes(status_as_u16(&self));
+        if let Some(description) = self.err().and_then(|e| e.description) {
+            res.extend(bool::into_bytes(true));
```

Comment:
This seems a bit overboard, but it seems that the presence / absence of the description could be packed into the first 2 bytes as well. Is a separate byte used for future compatibility (e.g., the desired serialization of `Option<T>`)?

Reply:
Packing this flag was my first intention, but I don't like code complication. I prefer to wait for support of enums in serialization and then update this implementation.